File: backend/Minimarket.Api/Services/SupplierSyncService.cs
using System.Globalization;
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using Minimarket.Api.DTOs;
using Minimarket.Api.Models;
using Minimarket.Api.Repositories;

namespace Minimarket.Api.Services;

// Sincroniza el catalogo del proveedor (Coca-Cola / AIC Digital - Arca Continental) hacia Productos.
// Reglas definidas en docs/PROJECT_CONTEXT.md (seccion de sincronizacion de catalogo de proveedor).
//
// IMPORTANTE: los nombres de las propiedades del JSON del proveedor (sku, longDescription, etc.)
// estan tomados de la investigacion documentada. Si la respuesta real difiere, ajustar las
// listas de candidatos en GetString/GetDecimal/GetInt y en FindArray (el parseo es tolerante).
public class SupplierSyncService(
    IHttpClientFactory httpClientFactory,
    IProductRepository productRepository,
    ICategoryRepository categoryRepository,
    ISupplierRepository supplierRepository,
    ISupplierProductRepository supplierProductRepository,
    ICompanyRepository companyRepository) : ISupplierSyncService
{
    private const int DefaultMinimumStock = 5;
    private const string CustomerId = "2898397";

    private const string CategoriesUrl =
        "https://briolightapimgmt.arcacontal.com/product/api/v1/Category/GetHomePageCategories?customerId=" + CustomerId;

    private const string PortfolioUrlFormat =
        "https://briolightapimgmt.arcacontal.com/product/api/v1/Portfolio?businessUnitId=4&CategoryId={0}" +
        "&Phrase=&Type=4&Limit=500&Offset=1&CustomerId=" + CustomerId + "&order=0&searchCriteria=";

    public async Task<(bool Success, string? Error, SupplierSyncResultDto? Result)> SyncAsync(SupplierSyncRequestDto request)
    {
        if (string.IsNullOrWhiteSpace(request.Token))
        {
            return (false, "Debes pegar el token del proveedor.", null);
        }

        if (string.IsNullOrWhiteSpace(request.SupplierDocumentNumber))
        {
            return (false, "Debes seleccionar el proveedor.", null);
        }

        var supplier = await supplierRepository.GetByDocumentNumberAsync(request.SupplierDocumentNumber.Trim());
        if (supplier is null)
        {
            return (false, "No se encontro un proveedor con ese numero de documento.", null);
        }

        var token = request.Token.Trim();
        var warnings = new List<string>();
        var categoriesProcessed = 0;
        var categoriesCreated = 0;
        var productsProcessed = 0;
        var productsCreated = 0;
        var productsUpdated = 0;

        var minimumStock = (await companyRepository.GetAsync())?.MinimumStock ?? DefaultMinimumStock;
        var expiration = DateOnly.FromDateTime(DateTime.Today.AddYears(2));
        var categoryCache = new Dictionary<string, Category>(StringComparer.OrdinalIgnoreCase);

        List<ProviderCategory> categories;
        try
        {
            categories = await FetchCategoriesAsync(token);
        }
        catch (SupplierSyncUnauthorizedException)
        {
            return (false, "El token esta vencido o no es valido. Vuelve a copiarlo del portal del proveedor.", null);
        }
        catch (HttpRequestException ex)
        {
            return (false, $"No se pudo conectar con el proveedor: {ex.Message}", null);
        }

        foreach (var category in categories)
        {
            categoriesProcessed++;
            var categoryName = string.IsNullOrWhiteSpace(category.Name) ? "Sin categoria" : category.Name.Trim();

            if (!categoryCache.TryGetValue(categoryName, out var localCategory))
            {
                localCategory = await categoryRepository.GetByNameAsync(categoryName);
                if (localCategory is null)
                {
                    categoriesCreated++;
                    if (!request.PreviewOnly)
                    {
                        localCategory = new Category { Name = Truncate(categoryName, 100), IsActive = true };
                        await categoryRepository.AddAsync(localCategory);
                        await categoryRepository.SaveChangesAsync();
                    }
                }

                if (localCategory is not null)
                {
                    categoryCache[categoryName] = localCategory;
                }
            }

            List<ProviderProduct> products;
            try
            {
                products = await FetchProductsAsync(token, category.ProviderId);
            }
            catch (SupplierSyncUnauthorizedException)
            {
                return (false, "El token esta vencido o no es valido. Vuelve a copiarlo del portal del proveedor.", null);
            }
            catch (HttpRequestException ex)
            {
                return (false, $"No se pudo conectar con el proveedor: {ex.Message}", null);
            }

            foreach (var providerProduct in products)
            {
                productsProcessed++;

                var units = providerProduct.Units;
                if (units <= 0)
                {
                    units = 1;
                    warnings.Add($"El producto '{Describe(providerProduct)}' no traia 'units' valido; se uso 1.");
                }

                // Costo: lo que paga el negocio al proveedor. NO se redondea a la decima (solo precision de 2 decimales).
                var cost = Math.Round(providerProduct.CustomerPrice / units, 2, MidpointRounding.AwayFromZero);
                // Precio de venta: se redondea a la decima mas cercana (regla de negocio).
                var price = Math.Round(providerProduct.SalePrice / units, 1, MidpointRounding.AwayFromZero);

                var name = Truncate(providerProduct.LongDescription, 150);
                if (string.IsNullOrWhiteSpace(name))
                {
                    name = Truncate(providerProduct.ShortDescription, 150);
                }

                if (string.IsNullOrWhiteSpace(name))
                {
                    warnings.Add($"Se omitio un producto sin nombre (sku '{providerProduct.Sku}').");
                    continue;
                }

                // El shortDescription del proveedor viene largo; armamos un nombre corto a partir del
                // formato del proveedor ("Nombre, empaque volumen, N Unidades").
                var shortName = BuildShortName(name);
                var sku = string.IsNullOrWhiteSpace(providerProduct.Sku) ? null : Truncate(providerProduct.Sku, 30);

                var existing = sku is null ? null : await productRepository.GetBySkuAsync(sku);

                if (existing is null)
                {
                    productsCreated++;
                    if (!request.PreviewOnly && localCategory is not null)
                    {
                        sku ??= await GenerateSkuAsync(categoryName);
                        var product = new Product
                        {
                            Name = name,
                            ShortName = shortName,
                            Sku = sku,
                            Barcode = null,
                            PurchaseBarcode = null,
                            Description = null,
                            Price = price,
                            Cost = cost,
                            Stock = 0,
                            MinimumStock = minimumStock,
                            ExpirationDate = expiration,
                            SalesUnitName = "Unidad",
                            PurchaseUnitName = "Unidad",
                            UnitsPerPurchaseUnit = 1,
                            IsActive = true,
                            CategoryId = localCategory.Id
                        };

                        await productRepository.AddAsync(product);
                        await productRepository.SaveChangesAsync();
                        await AddHistoryAsync(supplier.Id, product.Id, cost);
                    }
                }
                else
                {
                    productsUpdated++;
                    if (!request.PreviewOnly && localCategory is not null)
                    {
                        existing.Name = name;
                        existing.ShortName = shortName;
                        existing.Cost = cost;
                        existing.Price = price;
                        existing.CategoryId = localCategory.Id;
                        productRepository.Update(existing);
                        await productRepository.SaveChangesAsync();
                        await AddHistoryAsync(supplier.Id, existing.Id, cost);
                    }
                }
            }
        }

        var result = new SupplierSyncResultDto(
            request.PreviewOnly,
            supplier.Name,
            categoriesProcessed,
            categoriesCreated,
            productsProcessed,
            productsCreated,
            productsUpdated,
            warnings);

        return (true, null, result);
    }

    private async Task AddHistoryAsync(int supplierId, int productId, decimal cost)
    {
        await supplierProductRepository.AddAsync(new SupplierProduct
        {
            SupplierId = supplierId,
            ProductId = productId,
            LastCost = cost,
            Date = DateTime.Now
        });
        await supplierProductRepository.SaveChangesAsync();
    }

    private async Task<List<ProviderCategory>> FetchCategoriesAsync(string token)
    {
        var root = await GetJsonAsync(CategoriesUrl, token);
        var array = FindArray(root);
        var result = new List<ProviderCategory>();

        if (array.ValueKind != JsonValueKind.Array)
        {
            return result;
        }

        foreach (var element in array.EnumerateArray())
        {
            if (element.ValueKind != JsonValueKind.Object)
            {
                continue;
            }

            var id = GetInt(element, "id", "categoryId", "categoryID", "Id");
            var name = GetString(element, "name", "categoryName", "description", "title");
            if (id is not null)
            {
                result.Add(new ProviderCategory(id.Value, name ?? string.Empty));
            }
        }

        return result;
    }

    private async Task<List<ProviderProduct>> FetchProductsAsync(string token, int categoryId)
    {
        var url = string.Format(CultureInfo.InvariantCulture, PortfolioUrlFormat, categoryId);
        var root = await GetJsonAsync(url, token);
        var array = FindArray(root);
        var result = new List<ProviderProduct>();

        if (array.ValueKind != JsonValueKind.Array)
        {
            return result;
        }

        foreach (var element in array.EnumerateArray())
        {
            if (element.ValueKind != JsonValueKind.Object)
            {
                continue;
            }

            result.Add(new ProviderProduct(
                GetString(element, "sku", "SKU", "code") ?? string.Empty,
                GetString(element, "longDescription", "name", "description") ?? string.Empty,
                GetString(element, "shortDescription", "shortName") ?? string.Empty,
                GetDecimal(element, "salePrice", "price") ?? 0m,
                GetDecimal(element, "customerPrice", "cost") ?? 0m,
                GetDecimal(element, "units", "unit", "quantity") ?? 1m,
                GetInt(element, "categoryId", "categoryID", "category") ?? categoryId));
        }

        return result;
    }

    private async Task<JsonElement> GetJsonAsync(string url, string token)
    {
        var client = httpClientFactory.CreateClient();
        client.Timeout = TimeSpan.FromSeconds(90);

        using var requestMessage = new HttpRequestMessage(HttpMethod.Get, url);
        requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        using var response = await client.SendAsync(requestMessage);
        if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
        {
            throw new SupplierSyncUnauthorizedException();
        }

        response.EnsureSuccessStatusCode();

        await using var stream = await response.Content.ReadAsStreamAsync();
        using var document = await JsonDocument.ParseAsync(stream);
        return document.RootElement.Clone();
    }

    private async Task<string> GenerateSkuAsync(string categoryName)
    {
        var prefix = BuildCategoryPrefix(categoryName);
        var existing = await productRepository.GetSkusByPrefixAsync($"{prefix}-");
        var next = existing
            .Select(ParseSkuSequence)
            .DefaultIfEmpty(0)
            .Max() + 1;

        return $"{prefix}-{next:000000}";
    }

    private static int ParseSkuSequence(string sku)
    {
        var parts = sku.Split('-', 2);
        return parts.Length == 2 && int.TryParse(parts[1], out var value) ? value : 0;
    }

    private static string BuildCategoryPrefix(string categoryName)
    {
        var normalized = categoryName.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder();

        foreach (var character in normalized)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(character) == UnicodeCategory.NonSpacingMark)
            {
                continue;
            }

            if (char.IsLetterOrDigit(character))
            {
                builder.Append(char.ToUpperInvariant(character));
            }

            if (builder.Length == 3)
            {
                break;
            }
        }

        if (builder.Length == 0)
        {
            builder.Append("CAT");
        }

        while (builder.Length < 3)
        {
            builder.Append('X');
        }

        return builder.ToString();
    }

    private static string Truncate(string? value, int maxLength)
    {
        var trimmed = value?.Trim() ?? string.Empty;
        return trimmed.Length <= maxLength ? trimmed : trimmed[..maxLength];
    }

    private static readonly Regex VolumeRegex =
        new(@"\d+([.,]\d+)?\s*(ml|lt|l|kg|g)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    // Nombre corto a partir del formato del proveedor: toma el texto antes de la primera coma
    // (el nombre del producto) y le agrega el volumen (ml/L/g/kg) si no estaba ya incluido.
    // Asi se descartan el empaque ("PET", "Lata", "Vidrio...") y la cola "N Unidades".
    private static string BuildShortName(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return string.Empty;
        }

        var firstSegment = Regex.Replace(name.Split(',')[0].Trim(), @"\s+", " ");
        var match = VolumeRegex.Match(name);
        var volume = match.Success ? Regex.Replace(match.Value, @"\s+", string.Empty) : string.Empty;

        var result = firstSegment;
        if (volume.Length > 0 &&
            !firstSegment.Replace(" ", string.Empty).Contains(volume, StringComparison.OrdinalIgnoreCase))
        {
            result = $"{firstSegment} {volume}";
        }

        result = Regex.Replace(result, @"\s+", " ").Trim();
        return result.Length > 60 ? result[..60] : result;
    }

    private static string Describe(ProviderProduct product) =>
        !string.IsNullOrWhiteSpace(product.LongDescription) ? product.LongDescription
        : !string.IsNullOrWhiteSpace(product.Sku) ? product.Sku
        : "(sin nombre)";

    // Busca el array de datos dentro de la respuesta, sin importar el nombre del envoltorio.
    private static JsonElement FindArray(JsonElement root)
    {
        if (root.ValueKind == JsonValueKind.Array)
        {
            return root;
        }

        if (root.ValueKind != JsonValueKind.Object)
        {
            return default;
        }

        foreach (var key in new[] { "data", "items", "products", "result", "results", "portfolio", "categories", "value", "content" })
        {
            if (TryGetProperty(root, key, out var value) && value.ValueKind == JsonValueKind.Array)
            {
                return value;
            }
        }

        foreach (var property in root.EnumerateObject())
        {
            if (property.Value.ValueKind == JsonValueKind.Array)
            {
                return property.Value;
            }
        }

        foreach (var property in root.EnumerateObject())
        {
            if (property.Value.ValueKind == JsonValueKind.Object)
            {
                var nested = FindArray(property.Value);
                if (nested.ValueKind == JsonValueKind.Array)
                {
                    return nested;
                }
            }
        }

        return default;
    }

    private static bool TryGetProperty(JsonElement element, string name, out JsonElement value)
    {
        foreach (var property in element.EnumerateObject())
        {
            if (string.Equals(property.Name, name, StringComparison.OrdinalIgnoreCase))
            {
                value = property.Value;
                return true;
            }
        }

        value = default;
        return false;
    }

    private static string? GetString(JsonElement element, params string[] names)
    {
        foreach (var name in names)
        {
            if (TryGetProperty(element, name, out var value))
            {
                if (value.ValueKind == JsonValueKind.String)
                {
                    return value.GetString();
                }

                if (value.ValueKind is JsonValueKind.Number or JsonValueKind.True or JsonValueKind.False)
                {
                    return value.ToString();
                }
            }
        }

        return null;
    }

    private static decimal? GetDecimal(JsonElement element, params string[] names)
    {
        foreach (var name in names)
        {
            if (TryGetProperty(element, name, out var value))
            {
                if (value.ValueKind == JsonValueKind.Number && value.TryGetDecimal(out var number))
                {
                    return number;
                }

                if (value.ValueKind == JsonValueKind.String &&
                    decimal.TryParse(value.GetString(), NumberStyles.Any, CultureInfo.InvariantCulture, out var parsed))
                {
                    return parsed;
                }
            }
        }

        return null;
    }

    private static int? GetInt(JsonElement element, params string[] names)
    {
        var value = GetDecimal(element, names);
        return value is null ? null : (int)value.Value;
    }

    private sealed record ProviderCategory(int ProviderId, string Name);

    private sealed record ProviderProduct(
        string Sku,
        string LongDescription,
        string ShortDescription,
        decimal SalePrice,
        decimal CustomerPrice,
        decimal Units,
        int CategoryId);

    private sealed class SupplierSyncUnauthorizedException : Exception;
}

File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/service/SupplierSyncService.java
package com.minimarket.api.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.minimarket.api.dto.SupplierSyncRequestDto;
import com.minimarket.api.dto.SupplierSyncResultDto;
import com.minimarket.api.entity.Category;
import com.minimarket.api.entity.Product;
import com.minimarket.api.entity.Supplier;
import com.minimarket.api.entity.SupplierProduct;
import com.minimarket.api.repository.CategoryRepository;
import com.minimarket.api.repository.CompanyRepository;
import com.minimarket.api.repository.ProductRepository;
import com.minimarket.api.repository.SupplierProductRepository;
import com.minimarket.api.repository.SupplierRepository;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.text.Normalizer;
import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

// Sincroniza el catalogo del proveedor (Coca-Cola / AIC Digital - Arca Continental) hacia Productos.
// Reglas definidas en docs/PROJECT_CONTEXT.md (seccion de sincronizacion de catalogo de proveedor).
//
// IMPORTANTE: los nombres de las propiedades del JSON del proveedor (sku, longDescription, etc.)
// estan tomados de la investigacion documentada. Si la respuesta real difiere, ajustar las
// listas de candidatos en getString/getDecimal/getInt y en findArray (el parseo es tolerante).
@Service
public class SupplierSyncService {

    private static final int DEFAULT_MINIMUM_STOCK = 5;
    private static final String CUSTOMER_ID = "2898397";

    private static final String CATEGORIES_URL =
        "https://briolightapimgmt.arcacontal.com/product/api/v1/Category/GetHomePageCategories?customerId=" + CUSTOMER_ID;

    private static final String PORTFOLIO_URL_FORMAT =
        "https://briolightapimgmt.arcacontal.com/product/api/v1/Portfolio?businessUnitId=4&CategoryId=%d"
        + "&Phrase=&Type=4&Limit=500&Offset=1&CustomerId=" + CUSTOMER_ID + "&order=0&searchCriteria=";

    private final ProductRepository productRepository;
    private final CategoryRepository categoryRepository;
    private final SupplierRepository supplierRepository;
    private final SupplierProductRepository supplierProductRepository;
    private final CompanyRepository companyRepository;

    private final ObjectMapper objectMapper = new ObjectMapper();
    private final HttpClient httpClient = HttpClient.newHttpClient();

    public SupplierSyncService(ProductRepository productRepository,
                               CategoryRepository categoryRepository,
                               SupplierRepository supplierRepository,
                               SupplierProductRepository supplierProductRepository,
                               CompanyRepository companyRepository) {
        this.productRepository = productRepository;
        this.categoryRepository = categoryRepository;
        this.supplierRepository = supplierRepository;
        this.supplierProductRepository = supplierProductRepository;
        this.companyRepository = companyRepository;
    }

    public ServiceResult<SupplierSyncResultDto> sync(SupplierSyncRequestDto request) {
        if (request.token() == null || request.token().isBlank()) {
            return ServiceResult.failure("Debes pegar el token del proveedor.");
        }

        if (request.supplierDocumentNumber() == null || request.supplierDocumentNumber().isBlank()) {
            return ServiceResult.failure("Debes seleccionar el proveedor.");
        }

        var supplier = supplierRepository.findByDocumentNumber(request.supplierDocumentNumber().trim()).orElse(null);
        if (supplier == null) {
            return ServiceResult.failure("No se encontro un proveedor con ese numero de documento.");
        }

        var previewOnly = Boolean.TRUE.equals(request.previewOnly());
        var token = request.token().trim();
        var warnings = new ArrayList<String>();
        var counters = new Counters();

        var minimumStock = companyRepository.findById(1)
            .map(com.minimarket.api.entity.Company::getMinimumStock)
            .orElse(DEFAULT_MINIMUM_STOCK);
        var expiration = LocalDate.now().plusYears(2);
        var categoryCache = new HashMap<String, Category>();

        List<ProviderCategory> categories;
        try {
            categories = fetchCategories(token);
        } catch (UnauthorizedException ex) {
            return ServiceResult.failure("El token esta vencido o no es valido. Vuelve a copiarlo del portal del proveedor.");
        } catch (IOException | InterruptedException ex) {
            if (ex instanceof InterruptedException) {
                Thread.currentThread().interrupt();
            }
            return ServiceResult.failure("No se pudo conectar con el proveedor: " + ex.getMessage());
        }

        for (var category : categories) {
            counters.categoriesProcessed++;
            var categoryName = category.name() == null || category.name().isBlank() ? "Sin categoria" : category.name().trim();

            var localCategory = categoryCache.get(categoryName.toLowerCase());
            if (localCategory == null) {
                localCategory = categoryRepository.findByNameIgnoreCase(categoryName).orElse(null);
                if (localCategory == null) {
                    counters.categoriesCreated++;
                    if (!previewOnly) {
                        var nueva = new Category();
                        nueva.setName(truncate(categoryName, 100));
                        nueva.setIsActive(true);
                        localCategory = categoryRepository.save(nueva);
                    }
                }
                if (localCategory != null) {
                    categoryCache.put(categoryName.toLowerCase(), localCategory);
                }
            }

            List<ProviderProduct> products;
            try {
                products = fetchProducts(token, category.providerId());
            } catch (UnauthorizedException ex) {
                return ServiceResult.failure("El token esta vencido o no es valido. Vuelve a copiarlo del portal del proveedor.");
            } catch (IOException | InterruptedException ex) {
                if (ex instanceof InterruptedException) {
                    Thread.currentThread().interrupt();
                }
                return ServiceResult.failure("No se pudo conectar con el proveedor: " + ex.getMessage());
            }

            for (var providerProduct : products) {
                counters.productsProcessed++;

                var units = providerProduct.units();
                if (units == null || units.signum() <= 0) {
                    units = BigDecimal.ONE;
                    warnings.add("El producto '" + describe(providerProduct) + "' no traia 'units' valido; se uso 1.");
                }

                // Costo: lo que paga el negocio al proveedor. NO se redondea a la decima (solo 2 decimales de precision).
                var cost = providerProduct.customerPrice().divide(units, 2, RoundingMode.HALF_UP);
                // Precio de venta: se redondea a la decima mas cercana (regla de negocio).
                var price = providerProduct.salePrice().divide(units, 1, RoundingMode.HALF_UP);

                var name = truncate(providerProduct.longDescription(), 150);
                if (name.isBlank()) {
                    name = truncate(providerProduct.shortDescription(), 150);
                }

                if (name.isBlank()) {
                    warnings.add("Se omitio un producto sin nombre (sku '" + providerProduct.sku() + "').");
                    continue;
                }

                var shortName = truncate(providerProduct.shortDescription(), 60);
                var sku = providerProduct.sku() == null || providerProduct.sku().isBlank()
                    ? null
                    : truncate(providerProduct.sku(), 30);

                var existing = sku == null ? null : productRepository.findWithCategoryBySku(sku).orElse(null);

                if (existing == null) {
                    counters.productsCreated++;
                    if (!previewOnly && localCategory != null) {
                        if (sku == null) {
                            sku = generateSku(categoryName);
                        }
                        var product = new Product();
                        product.setName(name);
                        product.setShortName(shortName);
                        product.setSku(sku);
                        product.setBarcode(null);
                        product.setPurchaseBarcode(null);
                        product.setDescription(null);
                        product.setPrice(price);
                        product.setCost(cost);
                        product.setStock(0);
                        product.setMinimumStock(minimumStock);
                        product.setExpirationDate(expiration);
                        product.setSalesUnitName("Unidad");
                        product.setPurchaseUnitName("Unidad");
                        product.setUnitsPerPurchaseUnit(1);
                        product.setIsActive(true);
                        product.setCategoryId(localCategory.getId());

                        var saved = productRepository.save(product);
                        addHistory(supplier.getId(), saved.getId(), cost);
                    }
                } else {
                    counters.productsUpdated++;
                    if (!previewOnly && localCategory != null) {
                        existing.setName(name);
                        existing.setShortName(shortName);
                        existing.setCost(cost);
                        existing.setPrice(price);
                        existing.setCategoryId(localCategory.getId());
                        productRepository.save(existing);
                        addHistory(supplier.getId(), existing.getId(), cost);
                    }
                }
            }
        }

        var result = new SupplierSyncResultDto(
            previewOnly,
            supplier.getName(),
            counters.categoriesProcessed,
            counters.categoriesCreated,
            counters.productsProcessed,
            counters.productsCreated,
            counters.productsUpdated,
            warnings);

        return ServiceResult.success(result);
    }

    private void addHistory(Integer supplierId, Integer productId, BigDecimal cost) {
        var entry = new SupplierProduct();
        entry.setSupplierId(supplierId);
        entry.setProductId(productId);
        entry.setLastCost(cost);
        entry.setDate(LocalDateTime.now());
        supplierProductRepository.save(entry);
    }

    private List<ProviderCategory> fetchCategories(String token) throws IOException, InterruptedException {
        var root = getJson(CATEGORIES_URL, token);
        var array = findArray(root);
        var result = new ArrayList<ProviderCategory>();
        if (array == null || !array.isArray()) {
            return result;
        }

        for (var element : array) {
            if (!element.isObject()) {
                continue;
            }
            var id = getInt(element, "id", "categoryId", "categoryID", "Id");
            var name = getString(element, "name", "categoryName", "description", "title");
            if (id != null) {
                result.add(new ProviderCategory(id, name == null ? "" : name));
            }
        }

        return result;
    }

    private List<ProviderProduct> fetchProducts(String token, int categoryId) throws IOException, InterruptedException {
        var url = PORTFOLIO_URL_FORMAT.formatted(categoryId);
        var root = getJson(url, token);
        var array = findArray(root);
        var result = new ArrayList<ProviderProduct>();
        if (array == null || !array.isArray()) {
            return result;
        }

        for (var element : array) {
            if (!element.isObject()) {
                continue;
            }
            result.add(new ProviderProduct(
                orEmpty(getString(element, "sku", "SKU", "code")),
                orEmpty(getString(element, "longDescription", "name", "description")),
                orEmpty(getString(element, "shortDescription", "shortName")),
                orZero(getDecimal(element, "salePrice", "price")),
                orZero(getDecimal(element, "customerPrice", "cost")),
                orOne(getDecimal(element, "units", "unit", "quantity")),
                getInt(element, "categoryId", "categoryID", "category") == null
                    ? categoryId
                    : getInt(element, "categoryId", "categoryID", "category")));
        }

        return result;
    }

    private JsonNode getJson(String url, String token) throws IOException, InterruptedException {
        var httpRequest = HttpRequest.newBuilder(URI.create(url))
            .timeout(Duration.ofSeconds(90))
            .header("Authorization", "Bearer " + token)
            .header("Accept", "application/json")
            .GET()
            .build();

        var response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofByteArray());
        var status = response.statusCode();
        if (status == 401 || status == 403) {
            throw new UnauthorizedException();
        }
        if (status >= 400) {
            throw new IOException("HTTP " + status);
        }

        return objectMapper.readTree(response.body());
    }

    private String generateSku(String categoryName) {
        var prefix = buildCategoryPrefix(categoryName);
        var nextNumber = productRepository.findBySkuStartingWith(prefix + "-")
            .stream()
            .map(Product::getSku)
            .mapToInt(this::parseSkuSequence)
            .max()
            .orElse(0) + 1;

        return "%s-%06d".formatted(prefix, nextNumber);
    }

    private int parseSkuSequence(String sku) {
        var parts = sku.split("-", 2);
        if (parts.length != 2) {
            return 0;
        }
        try {
            return Integer.parseInt(parts[1]);
        } catch (NumberFormatException ex) {
            return 0;
        }
    }

    private String buildCategoryPrefix(String categoryName) {
        var normalized = Normalizer.normalize(categoryName, Normalizer.Form.NFD)
            .replaceAll("\\p{M}", "")
            .replaceAll("[^A-Za-z0-9]", "")
            .toUpperCase();

        if (normalized.isBlank()) {
            return "CAT";
        }

        var prefix = normalized.length() >= 3 ? normalized.substring(0, 3) : normalized;
        return (prefix + "XXX").substring(0, 3);
    }

    private static String describe(ProviderProduct product) {
        if (product.longDescription() != null && !product.longDescription().isBlank()) {
            return product.longDescription();
        }
        if (product.sku() != null && !product.sku().isBlank()) {
            return product.sku();
        }
        return "(sin nombre)";
    }

    private static String truncate(String value, int maxLength) {
        var trimmed = value == null ? "" : value.trim();
        return trimmed.length() <= maxLength ? trimmed : trimmed.substring(0, maxLength);
    }

    private static String orEmpty(String value) {
        return value == null ? "" : value;
    }

    private static BigDecimal orZero(BigDecimal value) {
        return value == null ? BigDecimal.ZERO : value;
    }

    private static BigDecimal orOne(BigDecimal value) {
        return value == null ? BigDecimal.ONE : value;
    }

    // Busca el array de datos dentro de la respuesta, sin importar el nombre del envoltorio.
    private static JsonNode findArray(JsonNode root) {
        if (root == null) {
            return null;
        }
        if (root.isArray()) {
            return root;
        }
        if (!root.isObject()) {
            return null;
        }

        for (var key : List.of("data", "items", "products", "result", "results", "portfolio", "categories", "value", "content")) {
            var value = field(root, key);
            if (value != null && value.isArray()) {
                return value;
            }
        }

        Iterator<JsonNode> elements = root.elements();
        while (elements.hasNext()) {
            var value = elements.next();
            if (value.isArray()) {
                return value;
            }
        }

        elements = root.elements();
        while (elements.hasNext()) {
            var value = elements.next();
            if (value.isObject()) {
                var nested = findArray(value);
                if (nested != null && nested.isArray()) {
                    return nested;
                }
            }
        }

        return null;
    }

    private static JsonNode field(JsonNode node, String name) {
        if (node == null || !node.isObject()) {
            return null;
        }
        Iterator<Map.Entry<String, JsonNode>> fields = node.fields();
        while (fields.hasNext()) {
            var entry = fields.next();
            if (entry.getKey().equalsIgnoreCase(name)) {
                return entry.getValue();
            }
        }
        return null;
    }

    private static String getString(JsonNode element, String... names) {
        for (var name : names) {
            var value = field(element, name);
            if (value != null && !value.isNull()) {
                return value.asText();
            }
        }
        return null;
    }

    private static BigDecimal getDecimal(JsonNode element, String... names) {
        for (var name : names) {
            var value = field(element, name);
            if (value == null || value.isNull()) {
                continue;
            }
            if (value.isNumber()) {
                return value.decimalValue();
            }
            if (value.isTextual()) {
                try {
                    return new BigDecimal(value.asText().trim());
                } catch (NumberFormatException ignored) {
                    // se ignora y se prueba el siguiente candidato
                }
            }
        }
        return null;
    }

    private static Integer getInt(JsonNode element, String... names) {
        var value = getDecimal(element, names);
        return value == null ? null : value.intValue();
    }

    private static final class Counters {
        int categoriesProcessed;
        int categoriesCreated;
        int productsProcessed;
        int productsCreated;
        int productsUpdated;
    }

    private record ProviderCategory(int providerId, String name) {
    }

    private record ProviderProduct(
        String sku,
        String longDescription,
        String shortDescription,
        BigDecimal salePrice,
        BigDecimal customerPrice,
        BigDecimal units,
        Integer categoryId) {
    }

    private static final class UnauthorizedException extends RuntimeException {
    }
}
